lib/net.js

```diff
@@ -452,10 +437,10 @@ function writeAfterFIN(chunk, encoding, cb) {
   // eslint-disable-next-line no-restricted-syntax
   const er = new Error('This socket has been ended by the other party');
   er.code = 'EPIPE';
-  process.nextTick(emitErrorNT, this, er);
   if (typeof cb === 'function') {
     defaultTriggerAsyncIdScope(this[async_id_symbol], process.nextTick, cb, er);
   }
+  errorOrDestroy(this, er, true);
```

Comment:
cb tick should be scheduled *before* emit error.

Reply:
It seems the `errorOrDestroy` utility is necessary for implementors. Can we avoid it somehow? Alternatively, we might want to consider moving it into the `Readable` and `Writable` prototypes and document it for implementors.

Reply:
I've been considering adding a `destroyMaybe` for implementors. I'll open a separate PR for that.

> Can we avoid it somehow? 

In this specific case we could allow the user providing a  `_writeAfterEnd` implementation which would be called. What `net` does here is a bit of monkey patching.